File: entrega1/atividades_posix/Java/Atividade1.java
import java.util.concurrent.Semaphore;
import java.util.Random;

class SharedBuffer {
    private final int[] buffer;
    private final int BUFFER_SIZE;
    private int in = 0;
    private int out = 0;
    
    private final Semaphore empty;
    private final Semaphore full;
    private final Object mutex = new Object();
    
    public SharedBuffer(int size) {
        this.BUFFER_SIZE = size;
        this.buffer = new int[BUFFER_SIZE];
        this.empty = new Semaphore(BUFFER_SIZE);
        this.full = new Semaphore(0);
    }
    
    public void produce(int item, int producerId) throws InterruptedException {
        empty.acquire();
        
        synchronized (mutex) {
            buffer[in] = item;
            System.out.printf("Produtor %d: produzido item %d na posição %d%n", 
                            producerId, item, in);
            in = (in + 1) % BUFFER_SIZE;
        }
        
        full.release();
    }
    
    public int consume(int consumerId) throws InterruptedException {
        full.acquire();
        
        int item;
        synchronized (mutex) {
            item = buffer[out];
            System.out.printf("Consumidor %d: consumido item %d da posição %d%n", 
                            consumerId, item, out);
            out = (out + 1) % BUFFER_SIZE;
        }
        
        empty.release();
        return item;
    }
}

class Producer extends Thread {
    private final SharedBuffer buffer;
    private final int producerId;
    private final int numItems;
    private final Random random = new Random();
    
    public Producer(SharedBuffer buffer, int id, int numItems) {
        this.buffer = buffer;
        this.producerId = id;
        this.numItems = numItems;
    }
    
    @Override
    public void run() {
        try {
            for (int i = 0; i < numItems; i++) {
                int item = random.nextInt(100) + 1;
                buffer.produce(item, producerId);
                Thread.sleep(random.nextInt(100));
            }
            System.out.printf("Produtor %d: finalizando%n", producerId);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

class Consumer extends Thread {
    private final SharedBuffer buffer;
    private final int consumerId;
    private final int numItems;
    private final Random random = new Random();
    
    public Consumer(SharedBuffer buffer, int id, int numItems) {
        this.buffer = buffer;
        this.consumerId = id;
        this.numItems = numItems;
    }
    
    @Override
    public void run() {
        try {
            for (int i = 0; i < numItems; i++) {
                buffer.consume(consumerId);
                Thread.sleep(random.nextInt(150));
            }
            System.out.printf("Consumidor %d: finalizando%n", consumerId);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

public class Atividade1 {
    private static final int BUFFER_SIZE = 10;
    private static final int NUM_ITEMS = 20;
    
    public static void main(String[] args) {
        int numProducers = 2;
        int numConsumers = 2;
        
        if (args.length == 2) {
            numProducers = Integer.parseInt(args[0]);
            numConsumers = Integer.parseInt(args[1]);
        }
        
        System.out.println("=== PRODUTOR/CONSUMIDOR - SEMÁFOROS ===");
        System.out.println("Buffer size: " + BUFFER_SIZE);
        System.out.println("Items per thread: " + NUM_ITEMS);
        System.out.println("Produtores: " + numProducers);
        System.out.println("Consumidores: " + numConsumers);
        System.out.println("========================================\n");
        
        SharedBuffer buffer = new SharedBuffer(BUFFER_SIZE);
        
        Producer[] producers = new Producer[numProducers];
        Consumer[] consumers = new Consumer[numConsumers];
        
        for (int i = 0; i < numProducers; i++) {
            producers[i] = new Producer(buffer, i + 1, NUM_ITEMS);
            producers[i].start();
        }
        
        for (int i = 0; i < numConsumers; i++) {
            consumers[i] = new Consumer(buffer, i + 1, NUM_ITEMS);
            consumers[i].start();
        }
        
        try {
            for (Producer producer : producers) {
                producer.join();
            }
            
            for (Consumer consumer : consumers) {
                consumer.join();
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        
        System.out.println("\n=== EXECUÇÃO FINALIZADA ===");
    }
}

File: entrega1/atividades_posix/C/atividade1.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>
#include <time.h>

#define BUFFER_SIZE 10
#define NUM_OPERATIONS 10

// Buffer compartilhado e suas variáveis de controle
int buffer[BUFFER_SIZE];
int in = 0, out = 0;

// Semáforos para sincronização
sem_t empty;  // Conta posições vazias
sem_t full;   // Conta posições ocupadas
sem_t mutex;  // Exclusão mútua para o buffer

// Estrutura para passar dados para as threads
typedef struct {
    int id;
    int num_operations;
} thread_data_t;

void *producer(void *arg) {
    thread_data_t *data = (thread_data_t *)arg;
    int producer_id = data->id;
    int operations = data->num_operations;
    
    for (int i = 0; i < operations; i++) {
        int item = rand() % 100 + 1; // Item entre 1 e 100
        
        // Esperar por slot vazio
        sem_wait(&empty);
        
        // Entrar na seção crítica
        sem_wait(&mutex);
        
        // Adicionar item ao buffer
        buffer[in] = item;
        printf("Produtor %d: produziu item %d na posição %d\n", 
               producer_id, item, in);
        in = (in + 1) % BUFFER_SIZE;
        
        // Sair da seção crítica
        sem_post(&mutex);
        
        // Sinalizar que há um item disponível
        sem_post(&full);
        
        // Simular tempo de produção
        usleep(rand() % 200000); // 0-200ms
    }
    
    printf("Produtor %d: finalizando\n", producer_id);
    return NULL;
}

void *consumer(void *arg) {
    thread_data_t *data = (thread_data_t *)arg;
    int consumer_id = data->id;
    int operations = data->num_operations;
    
    for (int i = 0; i < operations; i++) {
        // Esperar por item disponível
        sem_wait(&full);
        
        // Entrar na seção crítica
        sem_wait(&mutex);
        
        // Remover item do buffer
        int item = buffer[out];
        printf("Consumidor %d: consumiu item %d da posição %d\n", 
               consumer_id, item, out);
        out = (out + 1) % BUFFER_SIZE;
        
        // Sair da seção crítica
        sem_post(&mutex);
        
        // Sinalizar que há um slot vazio
        sem_post(&empty);
        
        // Simular tempo de consumo
        usleep(rand() % 300000); // 0-300ms
    }
    
    printf("Consumidor %d: finalizando\n", consumer_id);
    return NULL;
}

int main(int argc, char *argv[]) {
    int num_producers = 2;
    int num_consumers = 2;
    
    // Permitir configuração via argumentos
    if (argc == 3) {
        num_producers = atoi(argv[1]);
        num_consumers = atoi(argv[2]);
    }
    
    printf("=== PRODUTOR/CONSUMIDOR COM SEMÁFOROS ===\n");
    printf("Tamanho do buffer: %d\n", BUFFER_SIZE);
    printf("Operações por thread: %d\n", NUM_OPERATIONS);
    printf("Produtores: %d\n", num_producers);
    printf("Consumidores: %d\n", num_consumers);
    printf("==========================================\n\n");
    
    // Inicializar semáforos
    if (sem_init(&empty, 0, BUFFER_SIZE) != 0) {
        perror("Erro ao inicializar semáforo empty");
        exit(1);
    }
    
    if (sem_init(&full, 0, 0) != 0) {
        perror("Erro ao inicializar semáforo full");
        exit(1);
    }
    
    if (sem_init(&mutex, 0, 1) != 0) {
        perror("Erro ao inicializar semáforo mutex");
        exit(1);
    }
    
    // Inicializar gerador de números aleatórios
    srand(time(NULL));
    
    // Criar arrays de threads e dados
    pthread_t *producer_threads = malloc(num_producers * sizeof(pthread_t));
    pthread_t *consumer_threads = malloc(num_consumers * sizeof(pthread_t));
    thread_data_t *producer_data = malloc(num_producers * sizeof(thread_data_t));
    thread_data_t *consumer_data = malloc(num_consumers * sizeof(thread_data_t));
    
    clock_t start_time = clock();
    
    // Criar threads produtoras
    for (int i = 0; i < num_producers; i++) {
        producer_data[i].id = i + 1;
        producer_data[i].num_operations = NUM_OPERATIONS;
        
        if (pthread_create(&producer_threads[i], NULL, producer, &producer_data[i]) != 0) {
            perror("Erro ao criar thread produtora");
            exit(1);
        }
    }
    
    // Criar threads consumidoras
    for (int i = 0; i < num_consumers; i++) {
        consumer_data[i].id = i + 1;
        consumer_data[i].num_operations = NUM_OPERATIONS;
        
        if (pthread_create(&consumer_threads[i], NULL, consumer, &consumer_data[i]) != 0) {
            perror("Erro ao criar thread consumidora");
            exit(1);
        }
    }
    
    // Aguardar finalização das threads produtoras
    for (int i = 0; i < num_producers; i++) {
        if (pthread_join(producer_threads[i], NULL) != 0) {
            perror("Erro ao aguardar thread produtora");
            exit(1);
        }
    }
    
    // Aguardar finalização das threads consumidoras
    for (int i = 0; i < num_consumers; i++) {
        if (pthread_join(consumer_threads[i], NULL) != 0) {
            perror("Erro ao aguardar thread consumidora");
            exit(1);
        }
    }
    
    clock_t end_time = clock();
    double execution_time = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;
    
    printf("\n=== EXECUÇÃO FINALIZADA ===\n");
    printf("Tempo total: %.2f segundos\n", execution_time);
    printf("Total de itens produzidos: %d\n", num_producers * NUM_OPERATIONS);
    printf("Total de itens consumidos: %d\n", num_consumers * NUM_OPERATIONS);
    
    // Limpar recursos
    sem_destroy(&empty);
    sem_destroy(&full);
    sem_destroy(&mutex);
    
    free(producer_threads);
    free(consumer_threads);
    free(producer_data);
    free(consumer_data);
    
    return 0;
}